Zend/zend_vm_def.h

```diff
@@ -4261,9 +4276,13 @@ ZEND_VM_C_LABEL(fcall_end):
 		}
 	}
 
+ZEND_VM_C_LABEL(out):
 	if (UNEXPECTED(ZEND_CALL_INFO(call) & ZEND_CALL_RELEASE_THIS)) {
 		OBJ_RELEASE(Z_OBJ(call->This));
 	}
+	else if (UNEXPECTED(ZEND_CALL_INFO(call) & ZEND_CALL_CLOSURE)) {
+		OBJ_RELEASE(ZEND_CLOSURE_OBJECT(call->func));
+	}
```

Comment:
The same should be necessary for `ZEND_DO_FCALL`. You can reproduce this by compiling with `--enable-zend-test`.

Reply:
Did you mean `ZEND_DO_FCALL_BY_NAME` (the line where you added the comment is already in `ZEND_DO_FCALL`)? Is it even be possible for `ZEND_DO_FCALL_BY_NAME` to be used for Closure calls?

Locally I'm always building with `--enable-zend-test`, but that was a good hint. I'm seeing a use-after-free with `zend_test.replace_zend_execute_ex=1`, due to the `OBJ_RELEASE()` here:

```
021+ ==523228==ERROR: AddressSanitizer: heap-use-after-free on address 0x613000004640 at pc 0x555a72d468a9 bp 0x7fff27308a30 sp 0x7fff27308a28
022+ READ of size 4 at 0x613000004640 thread T0
023+     #0 0x555a72d468a8 in zend_gc_refcount php-src/Zend/zend_types.h:1314:12
024+     #1 0x555a72d52cd8 in zval_refcount_p php-src/Zend/zend_types.h:1363:9
025+     #2 0x555a72d1ab4b in zval_call_destructor php-src/Zend/zend_execute_API.c:216:35
026+     #3 0x555a72f1529a in zend_hash_reverse_apply php-src/Zend/zend_hash.c:2221:13
027+     #4 0x555a72d1a181 in shutdown_destructors php-src/Zend/zend_execute_API.c:261:4
028+     #5 0x555a72e0a6b0 in zend_call_destructors php-src/Zend/zend.c:1303:3
029+     #6 0x555a727d1f35 in php_request_shutdown php-src/main/main.c:1863:3
030+     #7 0x555a73e56ce8 in do_cli php-src/sapi/cli/php_cli.c:1126:3
031+     #8 0x555a73e4e024 in main php-src/sapi/cli/php_cli.c:1330:18
032+     #9 0x7faac8a2814f in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
033+     #10 0x7faac8a28208 in __libc_start_main csu/../csu/libc-start.c:360:3
034+     #11 0x555a708033a4 in _start (php-src/sapi/cli/php+0x1c033a4) (BuildId: 5ca217e112ed93f0f06fb995efe0241219cdc012)
035+ 
036+ 0x613000004640 is located 0 bytes inside of 328-byte region [0x613000004640,0x613000004788)
037+ freed by thread T0 here:
038+     #0 0x555a7089dd4a in free (php-src/sapi/cli/php+0x1c9dd4a) (BuildId: 5ca217e112ed93f0f06fb995efe0241219cdc012)
039+     #1 0x555a72bca845 in tracked_free php-src/Zend/zend_alloc.c:3015:2
040+     #2 0x555a72baf9a6 in _efree php-src/Zend/zend_alloc.c:2775:3
041+     #3 0x555a738f8168 in zend_objects_store_del php-src/Zend/zend_objects_API.c:204:3
042+     #4 0x555a736a4e1a in zend_object_release php-src/Zend/zend_objects_API.h:77:3
043+     #5 0x555a732a8b43 in ZEND_DO_FCALL_SPEC_RETVAL_UNUSED_HANDLER php-src/Zend/zend_vm_execute.h:1962:3
044+     #6 0x555a73002098 in execute_ex php-src/Zend/zend_vm_execute.h:57385:7
045+     #7 0x555a72758088 in custom_zend_execute_ex php-src/ext/zend_test/test.c:1097:2
046+     #8 0x555a730041c8 in zend_execute php-src/Zend/zend_vm_execute.h:63037:2
047+     #9 0x555a72e1bdfd in zend_execute_script php-src/Zend/zend.c:1906:3
048+     #10 0x555a727e5b38 in php_execute_script_ex php-src/main/main.c:2516:13
049+     #11 0x555a727e6be8 in php_execute_script php-src/main/main.c:2556:9
050+     #12 0x555a73e53934 in do_cli php-src/sapi/cli/php_cli.c:956:5
051+     #13 0x555a73e4e024 in main php-src/sapi/cli/php_cli.c:1330:18
052+     #14 0x7faac8a2814f in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
053+ 
054+ previously allocated by thread T0 here:
055+     #0 0x555a7089dff2 in malloc (php-src/sapi/cli/php+0x1c9dff2) (BuildId: 5ca217e112ed93f0f06fb995efe0241219cdc012)
056+     #1 0x555a72ba4c8a in tracked_malloc php-src/Zend/zend_alloc.c:2996:14
057+     #2 0x555a72baf404 in _emalloc php-src/Zend/zend_alloc.c:2765:10
058+     #3 0x555a737f1328 in zend_closure_new php-src/Zend/zend_closures.c:542:12
059+     #4 0x555a72e4da8c in _object_and_properties_init php-src/Zend/zend_API.c:1831:3
060+     #5 0x555a72e4ddd0 in object_init_ex php-src/Zend/zend_API.c:1845:9
061+     #6 0x555a737fa376 in zend_create_closure_ex php-src/Zend/zend_closures.c:736:2
062+     #7 0x555a737e8d5f in zend_create_closure php-src/Zend/zend_closures.c:837:2
063+     #8 0x555a734e1949 in ZEND_DECLARE_LAMBDA_FUNCTION_SPEC_CONST_HANDLER php-src/Zend/zend_vm_execute.h:5806:2
064+     #9 0x555a73002098 in execute_ex php-src/Zend/zend_vm_execute.h:57385:7
065+     #10 0x555a72758088 in custom_zend_execute_ex php-src/ext/zend_test/test.c:1097:2
066+     #11 0x555a730041c8 in zend_execute php-src/Zend/zend_vm_execute.h:63037:2
067+     #12 0x555a72e1bdfd in zend_execute_script php-src/Zend/zend.c:1906:3
068+     #13 0x555a727e5b38 in php_execute_script_ex php-src/main/main.c:2516:13
069+     #14 0x555a727e6be8 in php_execute_script php-src/main/main.c:2556:9
070+     #15 0x555a73e53934 in do_cli php-src/sapi/cli/php_cli.c:956:5
071+     #16 0x555a73e4e024 in main php-src/sapi/cli/php_cli.c:1330:18
072+     #17 0x7faac8a2814f in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
073+ 
```

so this line is apparently not quite correct.

Reply:
Oh, you are correct. `ZEND_DO_FCALL_BY_NAME` is what I meant, but it indeed is not used when the INIT code is `ZEND_INIT_DYNAMIC_CALL`. So, this should be a non-issue. Whatever you found is unrelated.

Reply:
Closure issue fixed in 6bf1e1b40549e9c9aa5aaac6bc8cd128285e3fbe.

Reply:
Okay, that commit added other issues. New attempt is here: c5942e98e682f8fb019b42405012a71362e35a46